lib/signers/pss_signer.dart

```diff
@@ -0,0 +1,275 @@
+// See file LICENSE for more information.
+
+library impl.signer.pss_signer;
+
+import 'dart:typed_data';
+
+import 'package:pointycastle/api.dart';
+import 'package:pointycastle/asymmetric/api.dart';
+import 'package:pointycastle/asymmetric/pkcs1.dart';
+import 'package:pointycastle/asymmetric/rsa.dart';
+import 'package:pointycastle/src/registry/registry.dart';
+
+import '../src/utils.dart';
+
+class PSSSigner implements Signer {
+  /// Intended for internal use.
+  static final FactoryConfig FACTORY_CONFIG =
+      DynamicFactoryConfig.suffix(Signer, '/RSA-PSS', (_, Match match) {
+    final digestName = match.group(1);
+    return () => PSSSigner(
+          PKCS1Encoding(RSAEngine()),
+          Digest(digestName),
+          Digest(digestName),
+        );
+  });
+
+  static const int TRAILER_IMPLICIT = 0xBC;
+
+  final Digest _contentDigest;
+  final Digest _mgfDigest;
+  final AsymmetricBlockCipher _cipher;
+
+  SecureRandom _random;
+
+  final int _hLen;
+  final int _mgfhLen;
+  final bool _sSet;
+  final int _sLen;
+  final Uint8List _mDash;
+  final int _trailer;
+
+  int _emBits;
+  Uint8List _salt;
+  Uint8List _block;
+
+  bool _forSigning;
+
+  PSSSigner(this._cipher, this._contentDigest, this._mgfDigest,
+      {int saltLength, int trailer = TRAILER_IMPLICIT})
+      : _hLen = _contentDigest.digestSize,
+        _mgfhLen = _mgfDigest.digestSize,
+        _sSet = false,
+        _sLen = saltLength,
+        _salt = Uint8List(saltLength),
+        _mDash = Uint8List(8 + saltLength + _contentDigest.digestSize),
+        _trailer = trailer;
+
+  PSSSigner.withSalt(this._cipher, this._contentDigest, this._mgfDigest,
```

Comment:
How should the signer be initialised with a salt? I'm thinking the correct way should to with the init() function or else people using the factory wouldn't be able to pass in their own salt. If this is the case, are there already subclasses of `CipherParameters` in the package which support salt configuration or do I need to implement my own? This is blocking the proper test implementations.

Reply:
From what I can tell there are 2 possible combinations of parameters users should be able to pass to `PSSSigner`, a `SecureRandom` with keys and salt length, and keys with salt.

For the case of keys with a salt the change should be simple and can be done like how `ParametersWithRandom` does it.

For the case of passing a `SecureRandom` with keys and a salt length I could add an implementation of `CipherParameters` with the salt config which itself contains a `ParametersWithRandom` object but that would mean users would have to go through 3 layers of nesting to get to the key objects which does not sound ideal.

Reply:
I opted to create two new parameter classes for the implementation, `ParametersWithSalt` and `ParametersWithSaltConfiguration`, tell me what you think.